### ISSUE TITLE
Encountered `process.env` when attempting to use `@wordpress/block-library` outside of WordPress

### ISSUE THREAD

**Issue by johnhooks** [WordPress/gutenberg#48949]
### Description

Why does `@wordpress/block-library` have a top level reference to `process.env`? I am trying to learn Gutenberg by playing with it outside of WordPress and using minimal build tools. Does `@wordpress/script` strip this out?

https://github.com/WordPress/gutenberg/blob/0722bc9e97389aa70da8f15030bd80bdccbe8d23/packages/block-library/src/index.js#L280-L295

### Step-by-step reproduction instructions

Use `registerCoreBlocks` of `@wordpress/block-library` in the browser without `@wordpress/scripts`, it will throw the following error:

```
Uncaught ReferenceError: process is not defined
```

### Environment info

```json
"@wordpress/block-editor": "11.5.0",
"@wordpress/block-library": "8.5.0",
"@wordpress/components": "23.5.0",
```

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by johnhooks** [comment]
I was able to replace `process.env.IS_GUTENBERG_PLUGIN` using `@rollup/plugin-replace` in my `vite.config.ts`. Though I still think referencing `process.env` is strange for a published package that is intended for use in the browser.

**Comment by talldan** [comment]
I don't think those of us that worked on it realised the side effects this would have for package consumers. It's been that way for a while and it was originally thought it would be okay as there was (and still is) use of `process.env.NODE_ENV` across the codebase.

The `IS_GUTENBERG_PLUGIN` property allows Gutenberg to ship stuff to the plugin, but exclude code from WordPress core, so it has an important purpose.

It'd be good to think about how this could be fixed.

**Comment by johnhooks** [comment]
I did find a solution, but if the intent is for the packages of Gutenberg to be usable outside the project a different approach for configuration would be more portable.

Any thoughts on what the state of `IS_GUTENBERG_PLUGIN` should be in this instance?

**Comment by talldan** [comment]
I guess one way to fix this is to ensure babel defines `process.env` when it builds packages.

> Any thoughts on what the state of IS_GUTENBERG_PLUGIN should be in this instance?

Usually `false`, but it depends what you're building. In this case setting it to `true` would define the `__experimentalRegisterExperimentalCoreBlocks` export for registering some of the more experimental blocks.

The best option may be to ensure `process.env` is defined as an empty object, and that way `IS_GUTENBERG_PLUGIN` will evaluate to `undefined` (and be falsey).

**Comment by johnhooks** [comment]
@talldan thanks for the suggestion. I intent to leave this issue open. I understand I am using the library totally outside the context it was created in, but there is value in publishing the packages in a way that is easier to use outside of Gutenberg and it's build system.

**Comment by johnhooks** [comment]
Decided this issue has a work around, it's documented here and people should be able to find it with a search, I'm closing it.

**Comment by gziolo** [comment]
> I was able to replace `process.env. IS_GUTENBERG_PLUGIN ` using `@rollup/plugin-replace` in my `vite.config.ts`. Though I still think referencing `process.env` is strange for a published package that is intended for use in the browser.

In retrospect, `process.env` wasn’t the best choice. In webpack 4, there was a polyfill for that env variable that projects overused. This variable should get replaced with something like `__IS_GUTENBERG_PLUGIN`. Still, it could have to be handled by the same webpack/rollup plugin. The benefit would be it wouldn’t error when not replaced.

There is a similar issue reported in https://github.com/WordPress/gutenberg/issues/44950.

**Comment by t-hamano** [comment]
I also recently encountered a similar error. That's when I tried to build a block editor for my React app by following [Getting Started](https://github.com/WordPress/gutenberg/blob/trunk/platform-docs/docs/intro.md) on platform-docs.

I used Vite as a build tool, but it looks like the latest Vite has removed `process.env` and I needed to use `import.meta.env` instead.

As mentioned in the comments below, I was able to work around this by add a `define` in `vite.config.js`.

https://github.com/vitejs/vite/issues/1973#issuecomment-787571499
